brigade.js

```diff
@@ -48,73 +37,138 @@ function build(e, project) {
-          })
-      }
+  return helmTest
+}
+
+// Here we can add additional Check Runs, which will run in parallel and
+// report their results independently to GitHub
+function runSuite(e, p) {
+  // Note: as provided language string is used in job naming, it must consist
+  // of lowercase letters and hyphens only (per Brigade/K8s restrictions)
+  checkRun(e, p, goTest, "go").catch(e  => {console.error(e.toString())});
+  checkRun(e, p, jsTest, "javascript").catch(e  => {console.error(e.toString())});
+  checkRun(e, p, helmTest, "helm").catch(e => {console.error(e.toString())});
+}
+
+// checkRun is a GitHub Check Run that is ran as part of a Checks Suite,
+// running the provided runFunc with language-based messaging
+function checkRun(e, p, runFunc, language) {
+  console.log("Check requested")
+
+  // Create Notification object (which is just a Job to update GH using the Checks API)
+  var note = new Notification(`test-${language}`, e, p);
+  note.conclusion = "";
+  note.title = `Run ${language} tests`;
+  note.summary = `Running the ${language} build/test targets for ${e.revision.commit}`;
+  note.text = "This task will ensure build, linting and tests all pass."
+
+  // Send notification, then run, then send pass/fail notification
+  return notificationWrap(runFunc(e, p), note)
+}
+
+// A GitHub Check Suite notification
+class Notification {
+  constructor(name, e, p) {
+    this.proj = p;
+    this.payload = e.payload;
+    this.name = name;
+    this.externalID = e.buildID;
+    this.detailsURL = `https://azure.github.io/kashti/builds/${ e.buildID }`;
+    this.title = "running check";
+    this.text = "";
+    this.summary = "";
+
+    // count allows us to send the notification multiple times, with a distinct pod name
+    // each time.
+    this.count = 0;
+
+    // One of: "success", "failure", "neutral", "cancelled", or "timed_out".
+    this.conclusion = "neutral";
+  }
+
+  // Send a new notification, and return a Promise<result>.
+  run() {
+    this.count++
+    var j = new Job(`${ this.name }-${ this.count }`, "deis/brigade-github-check-run:latest");
+    j.imageForcePull = true;
+    j.env = {
+      CHECK_CONCLUSION: this.conclusion,
+      CHECK_NAME: this.name,
+      CHECK_TITLE: this.title,
+      CHECK_PAYLOAD: this.payload,
+      CHECK_SUMMARY: this.summary,
+      CHECK_TEXT: this.text,
+      CHECK_DETAILS_URL: this.detailsURL,
+      CHECK_EXTERNAL_ID: this.externalID
     }
-    return Promise.resolve(runRelease)
-  }).catch(err => {
-    return ghNotify("failure", `failed build ${ e.buildID }`, e, project).run()
-  });
+    return j.run();
+  }
 }
 
-function releaseBrig(p, tag) {
+// Helper to wrap a job execution between two notifications.
+async function notificationWrap(job, note, conclusion) {
+  if (conclusion == null) {
+    conclusion = "success"
+  }
+  await note.run();
+  try {
+    let res = await job.run()
+    const logs = await job.logs();
+
+    note.conclusion = conclusion;
```

Comment:
maybe this could be set to 'success' directly

Reply:
👍 updated